x/concentrated-liquidity/pool_test.go

```diff
@@ -296,38 +294,204 @@ func (s *KeeperTestSuite) TestCalcOutAmtGivenIn() {
+
+				// add position two with the new price range above
+				_, _, _, err = s.App.ConcentratedLiquidityKeeper.CreatePosition(ctx, poolId, s.TestAccs[2], defaultAmt0, defaultAmt1, sdk.ZeroInt(), sdk.ZeroInt(), newLowerTick.Int64(), newUpperTick.Int64())
+				s.Require().NoError(err)
+			},
+			tokenIn:       sdk.NewCoin("usdc", sdk.NewInt(10000000000)),
+			tokenOutDenom: "eth",
+			priceLimit:    sdk.NewDec(6106),
+			// we expect to put 10000 usdc in and in return get 1.820536 eth back
+			// TODO: see why we don't get 9938.148 usdc and 1.80615 eth
+			expectedTokenOut: sdk.NewCoin("eth", sdk.NewInt(1820536)),
+			expectedTick:     sdk.NewInt(87173),
+			newLowerPrice:    sdk.NewDec(5501),
+			newUpperPrice:    sdk.NewDec(6250),
+		},
+	}
+
+	for name, test := range tests {
+		s.Run(name, func() {
+			s.Setup()
+			// create pool
+			pool, err := s.App.ConcentratedLiquidityKeeper.CreateNewConcentratedLiquidityPool(s.Ctx, 1, "eth", "usdc", currSqrtPrice, currTick)
+			s.Require().NoError(err)
+
+			// add positions
+			test.addPositions(s.Ctx, pool.Id)
+
+			// execute internal swap function
+			tokenOut, err := s.App.ConcentratedLiquidityKeeper.SwapOutAmtGivenIn(
+				s.Ctx,
+				test.tokenIn, test.tokenOutDenom,
+				swapFee, test.priceLimit, pool.Id)
+			s.Require().NoError(err)
+
+			pool = s.App.ConcentratedLiquidityKeeper.GetPoolbyId(s.Ctx, pool.Id)
+			s.Require().NoError(err)
+
+			// check that we produced the same token out from the swap function that we expected
+			s.Require().Equal(test.expectedTokenOut.String(), tokenOut.String())
+
+			// check that the pool's current tick was updated correctly
+			s.Require().Equal(test.expectedTick.String(), pool.CurrentTick.String())
+
+			// the following is needed to get the expected liquidity to later compare to what the pool was updated to
+			if test.newLowerPrice.IsNil() && test.newUpperPrice.IsNil() {
+				test.newLowerPrice = lowerPrice
+				test.newUpperPrice = upperPrice
+			}
+
+			newLowerTick := cl.PriceToTick(test.newLowerPrice)
+			newUpperTick := cl.PriceToTick(test.newUpperPrice)
+
+			lowerSqrtPrice, err := s.App.ConcentratedLiquidityKeeper.TickToSqrtPrice(newLowerTick)
+			s.Require().NoError(err)
+			upperSqrtPrice, err := s.App.ConcentratedLiquidityKeeper.TickToSqrtPrice(newUpperTick)
+			s.Require().NoError(err)
+
+			if test.poolLiqAmount0.IsNil() && test.poolLiqAmount1.IsNil() {
+				test.poolLiqAmount0 = defaultAmt0
+				test.poolLiqAmount1 = defaultAmt1
+			}
+
+			expectedLiquidity := cl.GetLiquidityFromAmounts(currSqrtPrice, lowerSqrtPrice, upperSqrtPrice, test.poolLiqAmount0, test.poolLiqAmount1)
+			// check that the pools liquidity was updated correctly
+			s.Require().Equal(expectedLiquidity.TruncateInt(), pool.Liquidity.TruncateInt())
+
+			// TODO: need to figure out a good way to test that the currentSqrtPrice that the pool is set to makes sense
+			// right now we calculate this value through iterations, so unsure how to do this here / if its needed
```

Comment:
Good call. Is there a trivial way to predict the number of iterations?

Please consider making an issue for this. I think it could be a good research task to tackle separately and asap

Reply:
https://github.com/osmosis-labs/osmosis/issues/3239

Will need to think more about best way to achieve this